### PR TITLE
Working energy usage for entity manipulation.

### PR DESCRIPTION
**Pull request by jagwire**
- Working example in examples/example/ui/MyEnergyBar.js

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
An admin must verify this patch for the pull request build to fire. Type "ok to test" to begin testing this pull request or "add to whitelist" to test this and all future pull requests from this collaborator.

**Comment by cozza13** [comment]
ok to test

**Comment by hifi-gustavo** [comment]
Build console output for commit: 47f7d55e2c5481f5c158beec2e91cdb7b211a27e available at: https://gist.github.com/bf2d27e4ebdcc5b28a9d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6997.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6997.dmg

**Comment by PhilipRosedale** [comment]
@murillodigital can you advise why this has failing checks?

**Comment by murillodigital** [comment]
Failed due to warnings:

[WARNINGS] Plug-in Result: Failed - 6 warnings exceed the threshold of 0 by 6

image

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7dd5bca17fad848eac97056fffa2e68a3ff61795 available at: https://gist.github.com/bf2d27e4ebdcc5b28a9d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6997.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6997.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 786400eb53c63fbed27d12bc6008768321c1032f available at: https://gist.github.com/bf2d27e4ebdcc5b28a9d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6997.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6997.dmg

File: interface/src/avatar/MyAvatar.cpp [hunk 1]
```diff
@@ -1851,3 +1863,31 @@ glm::mat4 MyAvatar::FollowHelper::postPhysicsUpdate(const MyAvatar& myAvatar, co
     }
 }
 
+float MyAvatar::getAccelerationEnergy() {
+    glm::vec3 velocity = getVelocity();
+    int changeInVelocity = abs(velocity.length() - priorVelocity.length());
+    float changeInEnergy = priorVelocity.length()*changeInVelocity*AVATAR_MOVEMENT_ENERGY_CONSTANT;
```
Comment (PhilipRosedale) [inline interface/src/avatar/MyAvatar.cpp]: spaces around * operator

File: interface/src/avatar/MyAvatar.cpp [hunk 2]
```diff
@@ -1851,3 +1863,31 @@ glm::mat4 MyAvatar::FollowHelper::postPhysicsUpdate(const MyAvatar& myAvatar, co
...
+float MyAvatar::getEnergy() {
+    return currentEnergy;
+}
+
+void MyAvatar::setEnergy(float value) {
+    currentEnergy = value;
+}
+
+float MyAvatar::getAudioEnergy() {
+    return getAudioLoudness()*AUDIO_ENERGY_CONSTANT;
```
Comment (PhilipRosedale) [inline interface/src/avatar/MyAvatar.cpp]: spaces around * operator

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 3]
```diff
@@ -122,6 +121,21 @@ QUuid EntityScriptingInterface::addEntity(const EntityItemProperties& properties
     EntityItemProperties propertiesWithSimID = convertLocationFromScriptSemantics(properties);
     propertiesWithSimID.setDimensionsInitialized(properties.dimensionsChanged());
 
+    auto dimensions = propertiesWithSimID.getDimensions();
+    float volume = dimensions.x*dimensions.y*dimensions.z;
+    auto density = propertiesWithSimID.getDensity();
+    auto newVelocity = propertiesWithSimID.getVelocity().length();
+    double cost = calculateCost(density*volume, 0, newVelocity);
```
Comment (PhilipRosedale) [inline libraries/entities/src/EntityScriptingInterface.cpp]: spaces around * operator

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 4]
```diff
@@ -211,6 +225,21 @@ EntityItemProperties EntityScriptingInterface::getEntityProperties(QUuid identit
 
 QUuid EntityScriptingInterface::editEntity(QUuid id, const EntityItemProperties& scriptSideProperties) {
     EntityItemProperties properties = scriptSideProperties;
+    
+    auto dimensions = properties.getDimensions();
+    float volume = dimensions.x*dimensions.y*dimensions.z;
+    auto density = properties.getDensity();
+    auto newVelocity = properties.getVelocity().length();
+    double cost = calculateCost(density*volume, 0, newVelocity);
```
Comment (PhilipRosedale) [inline libraries/entities/src/EntityScriptingInterface.cpp]: spaces around * operator

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 5]
```diff
@@ -211,6 +225,21 @@ EntityItemProperties EntityScriptingInterface::getEntityProperties(QUuid identit
 
 QUuid EntityScriptingInterface::editEntity(QUuid id, const EntityItemProperties& scriptSideProperties) {
     EntityItemProperties properties = scriptSideProperties;
+    
+    auto dimensions = properties.getDimensions();
+    float volume = dimensions.x*dimensions.y*dimensions.z;
```
Comment (PhilipRosedale) [inline libraries/entities/src/EntityScriptingInterface.cpp]: spaces around * operator

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 6]
```diff
@@ -316,6 +345,21 @@ void EntityScriptingInterface::deleteEntity(QUuid id) {
         _entityTree->withWriteLock([&] {
             EntityItemPointer entity = _entityTree->findEntityByEntityItemID(entityID);
             if (entity) {
+                
+                auto dimensions = entity->getDimensions();
+                float volume = dimensions.x*dimensions.y*dimensions.z;
```
Comment (PhilipRosedale) [inline libraries/entities/src/EntityScriptingInterface.cpp]: spaces around * operator

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 7]
```diff
@@ -992,3 +1036,20 @@ QStringList EntityScriptingInterface::getJointNames(const QUuid& entityID) {
                               Q_RETURN_ARG(QStringList, result), Q_ARG(QUuid, entityID));
     return result;
 }
+
+float EntityScriptingInterface::calculateCost(float mass, float oldVelocity,float newVelocity) {
```
Comment (PhilipRosedale) [inline libraries/entities/src/EntityScriptingInterface.cpp]: space after comma

File: libraries/entities/src/EntityScriptingInterface.cpp (hunk 4 above)
Comment (PhilipRosedale) [inline libraries/entities/src/EntityScriptingInterface.cpp]: This is an edit operation, so we need to check how much the velocity changed - the old velocity should go in, not zero.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0069849da3f310b5427abfed1a8cdedf1ef15618 available at: https://gist.github.com/bf2d27e4ebdcc5b28a9d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6997.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6997.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: aab98db812fb6c1ff14225619e687e564be899a5 available at: https://gist.github.com/bf2d27e4ebdcc5b28a9d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6997.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6997.dmg

**Comment by PhilipRosedale** [comment]
Works great!  Thanks.

**Pull request merged**